plugin/completion.py

```diff
@@ -1,25 +1,85 @@
 import sublime
 import sublime_plugin
-from .core.completion import parse_completion_response, format_completion
-from .core.configurations import is_supported_syntax
-from .core.documents import get_document_position, position_is_word
-from .core.edit import parse_text_edit
+
+from .core.protocol import Request, Range, InsertTextFormat
+from .core.settings import settings, client_configs
 from .core.logging import debug
-from .core.protocol import Request
+from .core.completion import parse_completion_response, format_completion
 from .core.registry import session_for_view, client_from_session, LSPViewEventListener
+from .core.configurations import is_supported_syntax
+from .core.documents import get_document_position
 from .core.sessions import Session
-from .core.settings import settings, client_configs
+from .core.edit import parse_text_edit
+from .core.views import range_to_region
 from .core.typing import Any, List, Dict, Tuple, Optional, Union
 
 
-class CompletionState(object):
-    IDLE = 0
-    REQUESTING = 1
-    APPLYING = 2
-    CANCELLING = 3
+last_text_command = None
 
 
-last_text_command = None
+class LspSelectCompletionItemCommand(sublime_plugin.TextCommand):
+    def run(self, edit: 'Any', item: 'Dict') -> None:
+        insert_text_format = item.get("insertTextFormat")
+
+        text_edit = item.get('textEdit')
+        if text_edit:
+            new_text = text_edit.get('newText')
+
+            range = Range.from_lsp(text_edit['range'])
+            edit_region = range_to_region(range, self.view)
+
+            if insert_text_format == InsertTextFormat.Snippet:
+                self.view.run_command("insert_snippet", {"contents": new_text})
+            else:
+                # use current point as the end of the edit region
+                current_point = self.view.sel()[0].begin()
+                edit_range = sublime.Region(edit_region.begin(), current_point)
```

Comment:
this is the only part of the PR that diverges from the spec.

I use the text edit, start ragne, 
but I can not use the text edit end range. 

This is because of the following reasons:
```
   | - is the curson

When choosing a command completion item,  1| pref|
                                          2|

                                          # removed
The typed prefix will be remove,          1| 
                                          2|

Because of the removed prefix, we cannot relay anymore on the range end position,
because the end position is outdated and it would select the wrong region.

We could potentionaly substract the prefix length from the end range.
But we cannot relay on that eather,
because sometimes the completion request is sent before the `textDocument/didChange`. 
(`textDocument/didChange` has also a debounce of 500ms);

I would not debounce the completion request for 500ms, because it feels really slow.

The only solution I found now is to use the current cursor postion. 
When choosing a command completion item,  1| pref|
                                          2|

The typed prefix will be remove,          1| 
                                            ^ this will be the end range position
                                          2|

and we will insert the completion item.   1| prefix|
                                          2|

I know this sounds complicated. But it is the only way I managed to make it work.
```